### PR TITLE
refactor dispense replicas for job completions

### LINKED ISSUE #2834: Scheduler: Assignment Algorithm Refactor TODO List
Now the assginment algorithm and code structure is complex, I'm trying to refactor the algorithms.

- [x] refactor assign replicas strategy #2827
- [x] refactor static weight dividing strategy #2845
- [x] refactor dynamic weight dividing strategy #2946
- [x] refactor aggregated dividing strategy #2946
- [ ] refactor job completion dividing algorithm #2964

### PR DESCRIPTION
**Pull request by Garrybest**
Signed-off-by: Garrybest <garrybest@foxmail.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #2834

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controllers/binding/common.go [hunk 1]
```diff
@@ -247,7 +247,7 @@ func divideReplicasByJobCompletions(workload *unstructured.Unstructured, cluster
 	}
 
 	if found {
-		targetClusters = util.DivideReplicasByTargetCluster(clusters, int32(completions))
```
Comment (jwcesign) [inline pkg/controllers/binding/common.go:250]: Can we delete this function? I found there is no place to invoke this

**Review by jwcesign — COMMENTED** [review]
Generally LGTM, just some nits

File: pkg/util/helper/binding.go [hunk 2]
```diff
@@ -68,6 +68,81 @@ func (p ClusterWeightInfoList) GetWeightSum() int64 {
...
 
+// Dispenser aims to divide replicas among clusters by different weights.
+type Dispenser struct {
+	// Target replicas, should be a positive integer.
+	NumReplicas int32
+	// Final result.
+	Result []workv1alpha2.TargetCluster
+}
+
+// NewDispenser will construct a dispenser will target replicas and a prescribed initial result.
```
Comment (jwcesign) [inline pkg/util/helper/binding.go:79]: NewDispenser will construct a dispenser with target replicas and a prescribed initial result?

**Review by jwcesign — COMMENTED** [review]
other lgtm

**Comment by jwcesign** [comment]
cc @RainbowMango for checking

**Review by RainbowMango — APPROVED** [review]
/approve

By the way, I still feel this part of the code is hard to understand.

**Pull request merged**